onnx_trt_backend.cpp

```diff
@@ -0,0 +1,771 @@
+#include "NvOnnxParser.h"
+#include "onnx/onnxifi.h"
+#include <NvInfer.h>
+#include <ctime>
+#include <thrust/device_vector.h>
+#include <unordered_map>
+
+namespace {
+
+struct InferDeleter {
+  template <typename T> void operator()(T *obj) const {
+    if (obj) {
+      obj->destroy();
+    }
+  }
+};
+template <typename T> inline std::shared_ptr<T> infer_object(T *obj) {
+  if (!obj) {
+    throw std::runtime_error("Failed to create object");
+  }
+  return std::shared_ptr<T>(obj, InferDeleter());
+}
+
+// Logger for GIE info/warning/errors
+class TRT_Logger : public nvinfer1::ILogger {
+  nvinfer1::ILogger::Severity _verbosity;
+  std::ostream *_ostream;
+
+public:
+  TRT_Logger(Severity verbosity = Severity::kWARNING,
+             std::ostream &ostream = std::cout)
+      : _verbosity(verbosity), _ostream(&ostream) {}
+  void log(Severity severity, const char *msg) override {
+    if (severity <= _verbosity) {
+      time_t rawtime = std::time(0);
+      char buf[256];
+      strftime(&buf[0], 256, "%Y-%m-%d %H:%M:%S", std::gmtime(&rawtime));
+      const char *sevstr =
+          (severity == Severity::kINTERNAL_ERROR
+               ? "    BUG"
+               : severity == Severity::kERROR
+                     ? "  ERROR"
+                     : severity == Severity::kWARNING
+                           ? "WARNING"
+                           : severity == Severity::kINFO ? "   INFO"
+                                                         : "UNKNOWN");
+      (*_ostream) << "[" << buf << " " << sevstr << "] " << msg << std::endl;
+    }
+  }
+};
+
+onnxStatus CheckShape(const nvinfer1::Dims &dims,
+                      const onnxTensorDescriptor &input) {
+  bool matched = false;
+  if (input.dimensions == dims.nbDims + 1) {
+    matched = true;
+    for (int i = 0; i < dims.nbDims; ++i) {
+      if (input.shape[i + 1] != dims.d[i]) {
+        matched = false;
```

Comment:
This function could be simpler if you just return early with `ONNXIFI_STATUS_MISMATCHING_SHAPE`